### ISSUE TITLE
New devfile based stacks should contain samples

### ISSUE THREAD

**Issue by l0rd** [eclipse-che/che#13529]
### Description

Currently a Che stack (i.e. `java-maven`) is associated to one or more samples (e.g. `web-java-petclinic` and `console-java-simple`). But often different samples of the same stack requires different settings (env variables, memory, commands) and making generic stacks that work with every project is hard.

Another point is that from a UX point of view it's more straightforward to select a `vue.js` or a `angular` stack rather than selecting the node js stack and then the angular sample.

So this issue is about the proposal to associate every stack to one unique project sample. The project, as well as the commands, would be specified in the the stack devfiles. The user would still be able to override the sample stack project and commands.

![image](https://user-images.githubusercontent.com/606959/59353098-8158dd80-8d22-11e9-8e21-909e36887746.png)

**Comment by l0rd** [comment]
| stack | sample |
| ----- | ------ |
| java-maven | https://github.com/che-samples/console-java-simple |
| java-gradle | https://github.com/che-samples/console-java-simple |
| java-spring-web | https://github.com/che-samples/web-java-spring.git |
| java-vertx | https://github.com/vert-x3/vertx-examples |
| java-postgresql | https://github.com/che-samples/web-java-spring-petclinic/tree/ |
| apache-camel | https://github.com/apache/camel/tree/master/examples/camel-example-spring-boot |
| python3 | https://github.com/che-samples/console-python3-simple.git |
| django | https://github.com/gothinkster/django-realworld-example-app |
| c# | https://github.com/che-samples/dotnet-web-simple.git |
| ASP.NET | https://github.com/che-samples/dotnet-web-simple.git |
| Go | https://github.com/golang/example/tree/master/outyet |
| Go | https://github.com/xesina/golang-echo-realworld-example-app |
| PHP | https://github.com/che-samples/web-php-simple |
| PHP-laravel | https://github.com/gothinkster/laravel-realworld-example-app |
| PHP-symfony | https://github.com/gothinkster/laravel-realworld-example-app |
| PHP-mysql | https://github.com/taniarascia/pdo/tree/master |
| nodejs | https://github.com/che-samples/web-nodejs-sample |
| nodejs-mongo | https://github.com/gothinkster/node-express-realworld-example-app |
| react | https://github.com/gothinkster/react-redux-realworld-example-app |
| angular | https://github.com/gothinkster/angular-realworld-example-app |
| vuejs | https://github.com/gothinkster/vue-realworld-example-app |

**Comment by skabashnyuk** [comment]
@l0rd is this issue https://github.com/eclipse/che/issues/13533 will be handled in your pr  https://github.com/eclipse/che-devfile-registry/pull/15 ? Do you know?

**Comment by l0rd** [comment]
@skabashnyuk that issue looked related to an outdated sample. The devfile I have added in the PR is not affected by that problem.

**Comment by l0rd** [comment]
Other nice to have stacks...

| stack |
| ------ |
| quarkus |  
| jib |  
| ko |  
| micronaut |
| microprofile |
| kubernetes-operator |
| che |  
| theia |  
| che-operator |  
| chectl |  
| che-plugin-broker |  
| vs-code-extension-dev |  
| che-doc |

**Comment by skabashnyuk** [comment]
@l0rd are they blocker for GA?

**Comment by l0rd** [comment]
"nice to have"

**Comment by metlos** [comment]
Before I leave on vacation, let me describe the state of the outstanding samples as I am aware about them.

## Vuejs

We can either wait for #13672 to be finished or, if we postpone that issue to 7.1 but need to have a vuejs sample in 7.0, we can use the following devfile in the sample (it will lack support for Vue-specific stuff but will have support for typescript/javascript and npm):

```yaml
---
apiVersion: 1.0.0
metadata:
  name: vuejs
projects:
  -
    name: vuejs-app
    source:
      type: git
      location: "https://github.com/gothinkster/vue-realworld-example-app.git"
components:
  -
    type: chePlugin
    id: che-incubator/typescript/latest
    memoryLimit: 512Mi
  -
    type: dockerimage
    alias: nodejs
    image: registry.access.redhat.com/ubi8/nodejs-10
    command: ['sleep']
    args: ['infinity']
    env:
      - name: HOME
        value: /tmp/user
      - name: PS1
        value: $(echo ${0})\\$
    memoryLimit: 512Mi
    endpoints:
      - name: 'vuejs-app'
        port: 8080
        attributes:
          discoverable: 'true'
    mountSources: true
commands:
  -
    name: run the web app
    actions:
      - type: exec
        component: nodejs
        command: npm install && npm run serve
        workdir: ${CHE_PROJECTS_ROOT}/vuejs-app
  -
    name: enable app preview
    actions:
      - type: exec
        component: nodejs
        command: |
          echo "module.exports = {devServer: {disableHostCheck: true}};" > vue.config.js
        workdir: ${CHE_PROJECTS_ROOT}/vuejs-app
```

## Go with https://github.com/xesina/golang-echo-realworld-example-app

This needs discussion with @tolusha and @tsmaeder on how best to define the sample, because our che plugin for Go uses golang 1.10.7 but the project requires Go modules, which are available from golang 1.11.

## java-mysql

I updated this devfile to use the oficial spring petclinic project.

## java-spring-web

I swapped the example for the spring-petclinic so that we have a more modern Spring-based application. As such this sample only differs from `java-mysql` in the fact that it doesn't defin the MySQL component and uses H2 instead.

##java-vertx

I think this is ready as is in the PR mentioned in the table.

**Comment by tolusha** [comment]
@l0rd @metlos @tsmaeder 
If nobody minds I will start working on `react` sample

**Comment by l0rd** [comment]
@tolusha go ahead

**Comment by sparkoo** [comment]
I've created an issue for Go realworld sample project #13803. This should be linked in the table https://github.com/eclipse/che/issues/13529#issuecomment-501746114. Could someone please edit the table and add reference there? Thanks

**Comment by sparkoo** [comment]
status update of https://github.com/eclipse/che/issues/13529#issuecomment-508229325
### Vuejs
postponed to 7.1.0 #13672
### java-spring-web & java-vertx & java-mysql
in review https://github.com/eclipse/che-devfile-registry/pull/18 which is blocked by #13796 (in review already https://github.com/eclipse/che-theia/pull/337)
### Go realworld
posponed to 7.1.0 #13803

**Comment by l0rd** [comment]
@sparkoo table updated

**Comment by tolusha** [comment]
@sparkoo 

c# and ASP.NET reference to the same project. Is it done by mistake? Because we already have the devfile for `dotnet-web-simple` [1]

c# | https://github.com/che-samples/dotnet-web-simple.git | PR
-- | -- | --
ASP.NET | https://github.com/che-samples/dotnet-web-simple.git | @tolusha

[1] https://github.com/eclipse/che-devfile-registry/blob/ab/fixPythonStack/devfiles/dotnet/devfile.yaml


c# | https://github.com/che-samples/dotnet-web-simple.git | 
-- | -- | --
ASP.NET | https://github.com/che-samples/dotnet-web-simple.git |

**Comment by sparkoo** [comment]
@l0rd @tolusha user experience in ASP.NET sample (in PR now https://github.com/eclipse/che-devfile-registry/pull/32) is not good. See my comments https://github.com/eclipse/che-devfile-registry/pull/32#issuecomment-510815810 and https://github.com/eclipse/che-devfile-registry/pull/32#issuecomment-511318510. It's probably caused by #13739 as @svor mentioned.
I would suggest to either fix #13739 or don't ship ASP.NET sample at all in 7.0.0GA. We should also retest C#.NET devfile as I guess that one will be also affected by that bug.

**Comment by sparkoo** [comment]
status update:
### java-spring-web & java-vertx & java-mysql
merged https://github.com/eclipse/che-devfile-registry/pull/18
### ASP.NET
in review https://github.com/eclipse/che-devfile-registry/pull/32. IMHO very bad user experience at this state. See my previous comment and PR itself.
### Go
New issue with `GOPATH` permissions #13823. There is a fix (https://github.com/eclipse/che-theia/pull/348), but it either doesn't fix the issue or it's not in `che-theia:next`. @tolusha is on it.
Note that this is about first, already merged, Go devfile!
### Vue.js and Go realworld
postponed to 7.1.0

---
I've created follow-up issue to retest all devfiles #13863

**Comment by skabashnyuk** [comment]
@tolusha what is your plan on https://github.com/eclipse/che-devfile-registry/pull/32 ?

**Comment by tolusha** [comment]
@skabashnyuk No

**Comment by sparkoo** [comment]
status update (@l0rd @skabashnyuk):
### ASP.NET
PR open https://github.com/eclipse/che-devfile-registry/pull/32. However, there is issue with Omnisharp plugin (https://github.com/eclipse/che/issues/13739) and without this fix, I vote to not include the ASP.NET sample to 7.0.0. This issue also affects C#.NET sample!
I would suggest to include #13739 into 7.0.0 and then merge ASP.NET XOR remove C#.NET devfile.

### Go
There is still issue #13823. I've testet just now with `eclipse/che-theia/7.0.0-next` and issue is still there. It is blocker for me as every go file with at least one external dependency won't work at this moment. I'm not sure whether the fix is wrong or it's just not in `7.0.0-next` image.

**Comment by l0rd** [comment]
I believe that the remaining problems are:

- The samples endpoints problems: https://github.com/eclipse/che/issues/13964
- Language servers not picking up new projects ([Java](https://github.com/eclipse/che/issues/13427) and [C# ](https://github.com/eclipse/che/issues/13739))

And we have dedicated issues for them. So I would close this issue @sparkoo is it ok for you?

**Comment by sparkoo** [comment]
@l0rd yes, I would close this one as well. We have all devfiles we want merged and following issues should be solved separately.
